### DISCUSSION TITLE
RFC: making control object optional for all the hooks

### DISCUSSION THREAD

**Discussion by bluebill1049** [react-hook-form/react-hook-form#7081] [RFC]
## Context

`control` object is a required `prop` for the following hooks. 

- `useController`
- `useFormState`
- `useFieldArray`
- `useWatch`

This is an internal object which provides a way for developers to avoid using `context` API (FormProvider).  However, it does make component integration a bit harder, especially when `control` prop becomes a required prop and will result in run time error without such. I would like to purpose `control` object to be optional, so developers can integrate with those hooks without worrying about `control` object being a hard dependency.

![Screen Shot 2021-11-19 at 8 21 37 pm](https://user-images.githubusercontent.com/10513364/142598140-f4a52c36-d431-49d1-99ab-74c9349c5ea5.png)
(Error screen above, when control is missing)

## Proposal

Provide a default value for the control prop internally, so avoid run time error when `control` object is missing not passed down.

### Props

- simplify the integration with react hook form's custom hooks

### Cons

- increase in package bundle size by providing default control object internally
- avoiding type check for `control` object, consider it's really an internal used object

## PR (draft)

https://github.com/react-hook-form/react-hook-form/pull/7046

cc @barrymay @kotarella1110 @jorisre @wdfinch @getTobiasNielsen

**Comment by barrymay** [comment]
To be honest, I think it's a nice-to-have, considering the cons - 1) I'm curious how you know which useForm its related to (i guess you would infer it from the parent or registered child?) and 2) I like the fact that I know exactly what form things are tied to.

**Comment by wdfinch** [comment]
I personally don't really have a need for this. It's easy to add the `control` prop especially if your fields are inside a `FormContextProvider` which automatically adds it. I guess if some people have an issue with adding this and it's not a ton of work to get around it might be better for some people (less config is better most of the time). I'd rather have the type checking personally.

**Comment by WalterWeidner** [comment]
I realize that this is a bit of a stale discussion but it is relevant to my use case. We have a pretty large internal component library and many of our custom form controls are controlled. We would love to leverage `useController` inside of these components to reduce boilerplate. However, if we do that we break all existing usage that isn't already inside of a FormProvider (which is probably most of the components) because `control` is required. 

It seems like this RFC would solve our problem. I don't think switching to uncontrolled components is feasible for us right now. Any recommendations on how to proceed?
